lib/Dialect/Moore/Transforms/LowerConcatRef.cpp

```diff
@@ -0,0 +1,126 @@
+
+namespace circt {
+namespace moore {
+#define GEN_PASS_DEF_LOWERCONCATREF
+#include "circt/Dialect/Moore/MoorePasses.h.inc"
+} // namespace moore
+} // namespace circt
+
+using namespace circt;
+using namespace moore;
+using namespace mlir;
+
+namespace {
+
+template <typename OpTy>
+struct ConcatRefLowering : public OpConversionPattern<ConcatRefOp> {
+  using OpConversionPattern::OpConversionPattern;
+
+  LogicalResult
+  matchAndRewrite(ConcatRefOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const final {
+    // Use to collect the operands of concatRef.
+    SmallVector<Value> values;
+    for (auto operand : op.getValues()) {
+      // Handle the nested concatRef, and then collect its operands.
+      if (auto *defOp = operand.getDefiningOp(); isa<ConcatRefOp>(defOp)) {
+        for (auto nestedOperand : defOp->getOperands())
+          values.push_back(nestedOperand);
+        rewriter.eraseOp(defOp);
+      } else
+        // If operands are not concatRef, directly collect operands.
+        values.push_back(operand);
+    }
+    // Create a new concatRef and replace all the old uses with the new ones.
+    // Then erase the old one.
+    auto newOp = rewriter.create<ConcatRefOp>(op->getLoc(),
+                                              op.getResult().getType(), values);
+    rewriter.replaceAllUsesWith(op.getResult(), newOp.getResult());
+    rewriter.eraseOp(op);
+
+    for (auto *user : newOp->getUsers())
+      // If the concatRef's user is an assignment, disassemble concatRef and
+      // create a new assignment using its operands with the different types.
+      if (isa<OpTy>(user)) {
+        auto src = user->getOperand(1);
+        auto srcWidth = cast<UnpackedType>(src.getType()).getBitSize().value();
+
+        for (auto operand : values) {
+          auto type = cast<RefType>(operand.getType()).getNestedType();
+          auto width = type.getBitSize().value();
+
+          rewriter.setInsertionPoint(user);
+          // FIXME: Need to estimate whether the bits range is from large to
+          // small or vice versa. Like "logic [7:0] or [0:7]".
+
+          // Only able to correctly handle the situation like "[7:0]" now.
+          auto i32 = moore::IntType::getInt(getContext(), 32);
+          auto lowBit = rewriter.create<ConstantOp>(newOp.getLoc(), i32,
+                                                    srcWidth - width);
+          auto extract =
+              rewriter.create<ExtractOp>(src.getLoc(), type, src, lowBit);
+
+          // Update the real bit width of RHS of assignment. Like "c" the above
+          // description mentioned.
+          srcWidth = srcWidth - width;
+          rewriter.create<OpTy>(user->getLoc(), operand, extract);
+        }
+
+        // Erase the old assignment if its LHS is concatRef.
+        rewriter.eraseOp(user);
+        // Because we needn't keep concatRef except it exists in instance.
+        rewriter.eraseOp(newOp);
```

Comment:
I think it must be after for loop?

Reply:
Yeah, you're right! I also think so! However, I don't know why print the same op three times. For example:
```
module {
  moore.module @top() {
    %a = moore.variable : <l5>
    %b = moore.variable : <l3>
    %c = moore.variable : <l8>
    %0 = moore.concat_ref %a, %b : (!moore.ref<l5>, !moore.ref<l3>) -> <l8>
    %1 = moore.read %c : l8
    moore.assign %0, %1 : l8
    moore.output
  }
}
```
When I add `op->dump()` at the start of the body of `matchAndRewrite()`, will dump
```
%0 = moore.concat_ref %a, %b : (!moore.ref<l5>, !moore.ref<l3>) -> <l8>
%0 = moore.concat_ref %a, %b : (!moore.ref<l5>, !moore.ref<l3>) -> <l8>
%0 = moore.concat_ref %a, %b : (!moore.ref<l5>, !moore.ref<l3>) -> <l8>
```
So when I move `rewriter.eraseOp(newOp);` after for loop, will throw an error:
```
failed to legalize operation 'moore.concat_ref' marked as erased
  assign {a,b} = c;
         ^
/home/phoenix/work/HDLBits/concatRef01.sv:5:10: note: see current operation: %3 = "moore.concat_ref"(%0, %1) : (!moore.ref<l5>, !moore.ref<l3>) -> !moore.ref<l8>
/home/phoenix/work/HDLBits/concatRef01.sv:5:10: note: found live user of result #0: moore.assign %0, %2 : l8
```
I'm not sure why this(dump three times) will occur.